django_cas_ng/views.py

```diff
@@ -33,12 +33,16 @@ def _service_url(request, redirect_to=None):
     service = urllib_parse.urlunparse(
         (protocol, host, request.get_full_path(), '', '', ''),
     )
+
+    # check if redirect param already in url
+    redirect_enc = urllib_parse.urlencode({REDIRECT_FIELD_NAME: redirect_to})
     if redirect_to:
-        if '?' in service:
-            service += '&'
-        else:
-            service += '?'
-        service += urllib_parse.urlencode({REDIRECT_FIELD_NAME: redirect_to})
```

Comment:
По задаче 2306 я должен был исправить проблему с юникодом в cas. После ее исправления запустил тесты (до исправления не запускал), тесты не проходили:

<pre>
tests/test_views_helpers.py ...F..........

========================================================================== FAILURES ==========================================================================
________________________________________________________ test_service_url_preserves_query_parameters _________________________________________________________
Traceback (most recent call last):
  File "/vagrant/venv/lib/python2.7/site-packages/django_cas_ng_repo/tests/test_views_helpers.py", line 53, in test_service_url_preserves_query_parameters
    assert 'foo=bar' in actual
AssertionError: assert 'foo=bar' in 'https://testserver/login/?next=https%3A%2F%2Ftestserver%2Flanding-page%2F'
</pre>

Ошибка не связана с моей правкой. Сейчас откатился на 59f567d - тесты не проходят, та же ошибка.
Вот что в логах:
```
commit 5eb8cc37fb94f1a0dce4045299a778c77861616e
Author: Jeffrey Gill <jpg18@case.edu>

    Fixes mingchen/django-cas-ng#17 "Forbidden in 3.4"

diff --git a/django_cas_ng/views.py b/django_cas_ng/views.py
--- a/django_cas_ng/views.py
+++ b/django_cas_ng/views.py
@@ -33,1 +33,1 @@
-        (protocol, host, request.get_full_path(), '', '', ''),
+        (protocol, host, request.path, '', '', ''),

commit c167264fc6d550216933cc406f1392823262dcab
Author: Ming Chen <mingchen@users.noreply.github.com>

    Revert 71d2a95(Removed the ticket GET param from the service, as it could break CAS.) to fix issue #17

diff --git a/django_cas_ng/views.py b/django_cas_ng/views.py
--- a/django_cas_ng/views.py
+++ b/django_cas_ng/views.py
@@ -41,1 +33,1 @@
-        (protocol, host, ticketless_full_path, '', '', ''),
+        (protocol, host, request.get_full_path(), '', '', ''),
```

Т.е. Jeffrey Gill <jpg18@case.edu> сломал тесты...

Смотрим тест:
```
def test_service_url_preserves_query_parameters():
    factory = RequestFactory()
    request = factory.get('/login/?foo=bar', secure=True)

    actual = _service_url(request, redirect_to='https://testserver/landing-page/')
    assert 'next=https%3A%2F%2Ftestserver%2Flanding-page%2F' in actual
    assert 'foo=bar' in actual
```
Падает в последней строке, параметр исходного запроса foo=bar не сохраняется (почему сейчас не происходит за двоение параметра)... Смотрим в код:
```
def _service_url(request, redirect_to=None):
    """Generates application service URL for CAS"""

    protocol = get_protocol(request)
    host = request.get_host()
    service = urllib_parse.urlunparse(
        (protocol, host, request.path, '', '', ''),
    )

...
    return service
```

Получается request = factory.get('/login/?foo=bar', secure=True), а service = 'https://testserver/login/', т.к. request.path = '/login/'... Что не правильно, т.к. параметры теряются. Исправляем на request.get_full_path() = '/login/?foo=bar'...

Получается этим исправлением ломаем 2 последних теста из #2328 (т.к. тесты скорее всего писались когда тесты cas были поломаны, поэтому проходили).

Первый "сломанный" тест из #2328 вообще не относится к cas, т.к. он написан был не правильно и просто рандомно падал.

Еще есть нюанс, у меня локально тесты себя странно ведут, многие из них не проходят (я писал об этом)...